File: src/main/java/at/ac/uibk/dps/cirrina/execution/service/ServiceImplementation.java
package at.ac.uibk.dps.cirrina.execution.service;

import at.ac.uibk.dps.cirrina.execution.object.context.ContextVariable;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Abstract service implementation, needs to be specialized.
 *
 * @see HttpServiceImplementation
 */
public abstract class ServiceImplementation {

  private final String name;
  private final boolean local;

  /**
   * Initializes this service implementation.
   *
   * @param name  Name of this service implementation.
   * @param local Whether this service implementation is local.
   */
  public ServiceImplementation(String name, boolean local) {
    this.name = name;
    this.local = local;
  }

  /**
   * Invoke this service implementation.
   *
   * @param input Input to the service invocation.
   * @param id    Sender ID.
   * @return The service invocation output.
   * @throws IOException If the service invocation failed.
   */
  public abstract CompletableFuture<List<ContextVariable>> invoke(
    List<ContextVariable> input,
    String id
  ) throws IOException;

  /**
   * Returns whether this service implementation is local.
   *
   * @return Is local.
   */
  public boolean isLocal() {
    return local;
  }

  /**
   * Returns the name of this service implementation.
   *
   * @return Name.
   */
  public String getName() {
    return name;
  }

  /**
   * Returns a string for informative purposes (a service implementation is abstract, so can provide no information about the details).
   *
   * @return Information string.
   */
  public abstract String getInformationString();
}

File: src/main/java/at/ac/uibk/dps/cirrina/execution/service/ServiceImplementation.kt
package at.ac.uibk.dps.cirrina.execution.service

import at.ac.uibk.dps.cirrina.execution.`object`.context.ContextVariable

/**
 * An abstract base class for service implementations within the execution engine.
 *
 * A service implementation defines how a specific type of service (e.g., HTTP, gRPC, or local) is
 * invoked. It handles the mapping of input variables to the service's native protocol and the
 * parsing of results back into context variables.
 *
 * @property name the unique name of the service implementation.
 * @property isLocal indicates whether the service is executed locally or remotely.
 */
abstract class ServiceImplementation(val name: String, val isLocal: Boolean) {

  /**
   * Invokes the service implementation asynchronously with the provided input.
   *
   * @param input the list of context variables to be passed to the service.
   * @return a [Result] containing the list of context variables returned by the service on success,
   *   or a failure if the invocation or response parsing fails.
   */
  abstract suspend fun invoke(input: List<ContextVariable>): Result<List<ContextVariable>>
}
